add support for matching multiple patterns

Uthar opened the pull request:
Added the ~`--and`~ ~`--expr`~ `--and` flags that search for matches containing all the patterns, similar to the `-and` flag of GNU find.

https://github.com/minad/consult/issues/6
#315

sharkdp commented:
Thank you very much for your contribution!

Before we consider the integration of this new feature, please see my comment in https://github.com/sharkdp/fd/issues/315#issuecomment-968323590:

> Ok, I'm inclined to accept a feature request to support `--and <pattern>`. Before we implement this, we need:
> 
> - A short discussion about the command-line option name. What do other tools use?
> - A detailed analysis if this could clash with any of the other command-line options or features of `fd`. There are some immediate questions like: what does `fd patternA --and patternB --type f` mean? (we are not going to support the meaning *patternA AND (patternB AND type==file)*).

Ideally, the second point should be discussed in detail and represented in corresponding unit tests.

Uthar commented:
> A short discussion about the command-line option name. What do other tools use?

- ripgrep uses `-e`

```
-e, --regexp <PATTERN>...                    
  A pattern to search for. This option can be provided multiple times, where
  all patterns given are searched. Lines matching at least one of the provided
  patterns are printed. This flag can also be used when searching for patterns
  that start with a dash.
  
  For example, to search for the literal '-foo', you can use this flag:
  
      rg -e -foo
  
  You can also use the special '--' delimiter to indicate that no more flags
  will be provided. Namely, the following is equivalent to the above:
  
      rg -- -foo

```
- GNU find uses the following syntax:

```
expr1 expr2
       Two expressions in a row are taken to be joined with an implied -a; expr2 is not evaluated if expr1 is false.

expr1 -a expr2
       Same as expr1 expr2.

expr1 -and expr2
       Same as expr1 expr2, but not POSIX compliant.
```

Uthar commented:
> what does fd patternA --and patternB --type f mean?

I think the `--and` flag would simply mean "Find all files whose names match all patterns specified by the initial [pattern] and all the --and flags"

No such nesting like you described would be possible

I don't know all of fd's flags, but it looks to me like it will compose well, being just an additional filter for names.

> unit tests

I created some basic unit tests. What other cases should be covered?

sharkdp commented:
> I think the `--and` flag would simply mean "Find all files whose names match all patterns specified by the initial [pattern] and all the --and flags"

Sounds reasonable. I would still like to see some thought going into potential conflicts (or chances for combination) with other command line options - like `--type` in the example quoted above. Other potentially interesting options to look at:
* `--glob`/`--fixed-strings`
* `--ignore-case`/`--case-sensitive`
* `--full-path`
* `--extension`/`--type`/`--size`/etc.


> a broken pattern in any of the --and flags should signal an error

That is definitely worth checking.

Another thing to potentially test would be patterns starting with a dash. Usually, we solve this using the `--` separator: `fd -- '-foo'`. Not sure if we could make it work for `--and` as well. Might not even require the `--` separator, but maybe a special clap setting to allow the value to start with a dash.

Uthar commented:
> patterns starting with a dash

One could also use `^[-]`

Uthar commented:
Hmm you're right, maybe an `--expr`, stolen from ripgrep, would sound better?

Uthar commented:
Wrote tests for all these in new commit:

> --glob

`--and` patterns contribute to the search as globs, not regexps


> --fixed-strings

`--and` patterns are matched as whole strings, not regexp


> --ignore-case

case is ignored in `--and` patterns as well as the primary pattern


> --case-sensitive

case is sensitive in `--and` patterns as well as the primary pattern


> --full-path

`--and` patterns contribute to the search in the whole path, not just file or dir name


> --extension

`--extension` and `--and` patterns are composed in a logical way, i.e. patterns match file names via AND and extension matches extension via OR


> --type

`--type` additionally restricts the search done by all the `--and` patterns 


> --size

didn't check, I didn't know how

also added these from my post:

> test with empty patterns
> test with unreasonable pattern should match nothing
> a broken pattern in any of the --and flags should signal an error


tests fails for non-ubuntu CI... let me check...

Uthar commented:
> tests fails for non-ubuntu CI... let me check...

Can't make sense out of it. Is there some Mac and Windows experts on board?

tmccombs commented:
We probably want to clearly document how `--and` interacts with other options. The behaviors described above makes sense to me and are probably what I would expect in the absence of documentation, but I could easily see someone else expecting different behavior. 

Regarding hyphens. We probably want to use the [`allow_hyphen_values`](https://docs.rs/clap/latest/clap/struct.Arg.html#method.allow_hyphen_values) option on the argument.

sharkdp commented:
@Uthar Thank you very much for the updates! Unfortunately, there are some larger merge conflicts now that I merged a few other PRs, especially #1067. I hope you can sort this out, otherwise please let us know.

Uthar commented:
Yes, I will attempt to solve this, stay tuned

Uthar commented:
Damn, looks like I will have to write it from scratch
It will take some time, I haven't written any rust code before

tmccombs commented:
I'm willing to help if you would like. After all, I made the PR that caused most of the conflicts.

Uthar commented:
Very nice of you tmccombs. My main problem is to write the code again, it takes a lot of time especially debugging the type errors with borrow and so on, because I lack experience.  I just don't have the time right now

Ultimately we just have to change `pattern` to `patterns` in this line:

https://github.com/sharkdp/fd/blob/master/src/walk.rs#L153

Everything else is just argument parsing/changing parameters to functions

Uthar commented:
I started some work here, the easy stuff :-) https://github.com/Uthar/fd/tree/conflicts

Uthar commented:
Ok, I've got this working now... but there are some issues that Rust experienced programmer will probably easily solve:

I'm trying to push the "initial" pattern to the optional vector of "and" patterns, but can't do that without borrowing, do you know how to handle it?

Also its panicking with no --expr flag currently, maybe unwrap() is not the best there?

I also don't know how to exit the program from the match .. Ok .. Err clause, when there is an error, it still requires a regex type.

Could you help me with these issues? See this commit: https://github.com/Uthar/fd/commit/d66470a50b05a1eb176d5ce37a2e6f89d32171a1

Uthar commented:
Full diff here https://github.com/sharkdp/fd/compare/master...Uthar:fd:conflicts

Uthar commented:
Done after all, we're back in the game :-)

But you should review the changes

sharkdp commented:
Thank you. The tests are currently failing on Windows and macOS

Uthar commented:
yeah... the failuress are really weird, dont make sense...

My friend has a macbook, I will ask them to try on it, also try to find a windows machine

Uthar commented:
fixed the tests

tmccombs commented on src/main.rs:
```diff
@@ -166,11 +192,14 @@ fn check_path_separator_length(path_separator: Option<&str>) -> Result<()> {
 }
 
-fn construct_config(mut opts: Opts, pattern_regex: &str) -> Result<Config> {
+fn construct_config(mut opts: Opts, pattern_regexps: &Vec<String>) -> Result<Config> {
```
This should probably be:

```suggestion
fn construct_config(mut opts: Opts, pattern_regexps: &[String]) -> Result<Config> {
```
(note that a `Vec<String>` will implicitly dereference to a `&[String]`, so `&pattern_regexps` would still fit the type)

We could almost get away with just accepting an iterator, except that we would need to be able to iterate over it it twice in order to determine if we should be case sensitive or not if the options aren't expllicit.

tmccombs commented on src/main.rs:
```diff
@@ -409,9 +438,13 @@ fn extract_time_constraints(opts: &Opts) -> Result<Vec<TimeFilter>> {
 fn ensure_use_hidden_option_for_leading_dot_pattern(
     config: &Config,
-    pattern_regex: &str,
+    pattern_regexps: &Vec<String>,
```
Similaraly this should be `&[String]` or maybe `&[&str]`

Uthar commented:
Thank you, tmccombs, for the thorough review.

See the last commit, where I added your solution to avoid `match` and `panic!` during generation of the regexp and changed type hints as you requested.

Let me know if you have some other improvement ideas.

sharkdp commented on src/cli.rs:
```diff
@@ -523,6 +523,15 @@ pub struct Opts {
+        value_name = "pattern",
+        long_help = "(Additionally) filter search results by all patterns via AND. Multiple \
+                     additionalpatterns can be specified."
+    )]
```
I would appreciate if we keep this out of the short help text for now (`hide_short_help = true`).

sharkdp commented on src/cli.rs:
```diff
@@ -523,6 +523,15 @@ pub struct Opts {
+    #[arg(
+        long = "expr",
+        value_name = "pattern",
+        long_help = "(Additionally) filter search results by all patterns via AND. Multiple \
+                     additionalpatterns can be specified."
```
Maybe
```suggestion
        long_help = "Add additional required search patterns, all of which must be fulfilled. Multiple \
                     additional patterns can be specified. The pattern is a regular expression, unless '--glob' \
                     or '--fixed-strings' is used."
```

sharkdp commented on src/cli.rs:
```diff
@@ -523,6 +523,15 @@ pub struct Opts {
     )]
     pub pattern: String,
 
+    /// Additional search patterns to match a'la AND (a regular expression, unless '--glob' is used; optional)
```
```suggestion
    /// Additional search patterns that need to be matched
```

sharkdp commented on src/main.rs:
```diff
@@ -81,12 +81,28 @@ fn run() -> Result<ExitCode> {
+
+    let pattern_regexps = exprs
+        .as_ref()
+        .unwrap_or(&empty)
```
```suggestion
        .unwrap_or_else(|| vec![])
```

sharkdp commented on src/main.rs:
```diff
@@ -409,9 +427,13 @@ fn extract_time_constraints(opts: &Opts) -> Result<Vec<TimeFilter>> {
+        && config.ignore_hidden
+        && pattern_regexps
+            .iter()
+            .any(|pat| pattern_matches_strings_with_leading_dot(pat))
```
does this work?

```suggestion
            .any(pattern_matches_strings_with_leading_dot)
```

Uthar replied:
I'm getting: `rustc [E0631]: expected signature of fn(&std::string::String) -> _`

sharkdp replied:
ok, then let's leave this.

sharkdp commented on src/main.rs:
```diff
@@ -166,11 +181,14 @@ fn check_path_separator_length(path_separator: Option<&str>) -> Result<()> {
     // if the pattern has an uppercase character (smart case).
-    let case_sensitive =
-        !opts.ignore_case && (opts.case_sensitive || pattern_has_uppercase_char(pattern_regex));
+    let case_sensitive = !opts.ignore_case
+        && (opts.case_sensitive
+            || pattern_regexps
+                .iter()
+                .any(|pat| pattern_has_uppercase_char(pat)));
```
please update the doc comment above (or if any of the patterns has ...)

sharkdp commented on src/main.rs:
```diff
@@ -81,12 +81,28 @@ fn run() -> Result<ExitCode> {
+
+    let regexps = pattern_regexps
+        .iter()
+        .map(|pat| build_regex(pat.to_string(), &config))
```
isn't `pat` a string already?

Uthar replied:
There is an error message when I remove the `to_string()`:

```
error[E0308]: mismatched types: expected struct `std::string::String`, found `&std::string::String
```

sharkdp replied:
Then I guess you could use `.into_iter()` instead of `.iter()` (and remove `.to_string()`)?

Uthar replied:
https://github.com/sharkdp/fd/compare/d8196a109e81e812ac9e558363a210ca2d660e11..1409ad6d306b400c07d75538a804810d8a569bc0

Uthar commented on src/walk.rs:
```diff
@@ -49,7 +49,7 @@ pub const DEFAULT_MAX_BUFFER_TIME: Duration = Duration::from_millis(100);
 /// If the `--exec` argument was supplied, this will create a thread pool for executing
```
pattern -> patterns

Uthar commented:
https://github.com/sharkdp/fd/blob/882e13a77276255e21c65af7eef6382a1fcb45a8/src/walk.rs#L47

pattern -> patterns

Uthar commented:
So the `--expr` flag instead of `--and` sounds better to you?

Uthar commented:
https://github.com/sharkdp/fd/blob/d498724edd8be4680ea74ad235f4d2ee8a304911/src/main.rs#L438-L442

The pattern -> A pattern / Some patterns ?

sharkdp commented:
> So the `--expr` flag instead of `--and` sounds better to you?

No, I like `--and`.

> The pattern -> A pattern / Some patterns ?

Maybe: "The pattern(s) seems to only match… or adjust your search pattern(s)".

sharkdp commented on src/cli.rs:
```diff
@@ -523,6 +523,17 @@ pub struct Opts {
+        value_name = "pattern",
+        long_help = "Add additional required search patterns, all of which must be fulfilled. Multiple \
+                     additional patterns can be specified. The patterns are regular expressions,
+                     unless '--glob'or '--fixed-strings' is used.",
```
```suggestion
                     unless '--glob' or '--fixed-strings' is used.",
```

sharkdp commented on src/main.rs:
```diff
@@ -166,11 +181,14 @@ fn check_path_separator_length(path_separator: Option<&str>) -> Result<()> {
-    // if the pattern has an uppercase character (smart case).
-    let case_sensitive =
-        !opts.ignore_case && (opts.case_sensitive || pattern_has_uppercase_char(pattern_regex));
+    // if any of the patterns have an uppercase character (smart case).
```
minor
```suggestion
    // if any of the patterns has an uppercase character (smart case).
```

Uthar replied:
https://github.com/sharkdp/fd/compare/5794d62b45d4db9d02a19b967ec3d3ad3171c36f..d8196a109e81e812ac9e558363a210ca2d660e11

sharkdp reviewed (approved):
Thank you. I made a couple of minor improvements.
